common/elasticsearch/esql/esql.go

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package esql
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+
+	"github.com/xwb1989/sqlparser"
+)
+
+// Replace ...
+// esql use replace function to apply user colName or column value replacing policy
+type Replace func(string) (string, error)
+
+// Filter ...
+// esql use filter function decide whether the policy will be applied to the column
+// only accept column names that filter(colName) == true
+type Filter func(string) bool
+
+// ESql ...
+// ESql is used to hold necessary information that required in parsing
+type ESql struct {
+	filterReplace Filter  // select the column we want to replace name
+	filterProcess Filter  // select the column we want to process value
+	replace       Replace // if selected by filterReplace, change the column name
+	process       Replace // if selected by filterProcess, change the column value
+	cadence       bool
+	pageSize      int
+	bucketNumber  int
+}
+
+// SetDefault ...
+// all members goes to default
+// should not be called if there is potential race condition
+func (e *ESql) SetDefault() {
+	e.pageSize = DefaultPageSize
+	e.bucketNumber = DefaultBucketNumber
+	e.cadence = false
+	e.replace = nil
+	e.process = nil
+	e.filterReplace = nil
+	e.filterProcess = nil
+}
+
+// NewESql ... return a new default ESql
+func NewESql() *ESql {
+	return &ESql{
+		pageSize:      DefaultPageSize,
+		bucketNumber:  DefaultBucketNumber,
+		cadence:       false,
+		process:       nil,
+		replace:       nil,
+		filterReplace: nil,
+		filterProcess: nil,
+	}
+}
+
+// SetReplace ... set up user specified column name replacement policy
+// should not be called if there is potential race condition
+func (e *ESql) SetReplace(filterArg Filter, replaceArg Replace) {
```

Comment:
Can we add support for multiple filters? Also let's discuss we may need rename to ProcessColumn and ProcessValue

Reply:
Sure.